db-service/lib/SQLService.js

```diff
@@ -27,62 +26,77 @@ class SQLService extends DatabaseService {
     this.on(['*'], this.onPlainSQL)
     return super.init()
   }
 
-  /** @type {Handler} */
-  async transformStreamFromCQN({ query }, next) {
-    if (!query._streaming) return next()
-    const cqn = STREAM.from(query.SELECT.from).column(query.SELECT.columns[0].ref[0])
-    if (query.SELECT.where) cqn.STREAM.where = query.SELECT.where
-    const stream = await this.run(cqn)
-    return stream && { value: stream }
-  }
-
-  /** @type {Handler} */
-  async transformStreamIntoCQN({ query, data, target }, next) {
-    let col, type, etag
-    const elements = query._target?.elements || target?.elements
-    if (!elements) next()
-    for (const key in elements) {
-      const element = elements[key]
-      if (element['@Core.MediaType'] && data[key]?.pipe) col = key
-      if (element['@Core.IsMediaType'] && data[key]) type = key
-      if (element['@odata.etag'] && data[key]) etag = key
-    }
-
-    if (!col) return next()
-
-    const cqn = STREAM.into(query.UPDATE.entity).column(col).data(data[col])
-    if (query.UPDATE.where) cqn.STREAM.where = query.UPDATE.where
-    const result = await this.run(cqn)
-    if (type || etag) {
-      const d = { ...data }
-      delete d[col]
-      const cqn = UPDATE.entity(query.UPDATE.entity).with(d)
-      if (query.UPDATE.where) cqn.UPDATE.where = query.UPDATE.where
-      await this.run(cqn)
+  _changeToStreams(columns, rows, one, compat) {
+    if (!rows || !columns) return
+    if (!Array.isArray(rows)) rows = [rows]
+    if (!rows.length || !Object.keys(rows[0]).length) return
+
+    // REVISIT: remove after removing compat_stream_cqn feature flag 
+    if (compat) { 
+      rows[0][Object.keys(rows[0])[0]] = this._stream(Object.values(rows[0])[0])
+      return
+    }   
+
+    for (let col of columns) {
+      const name = col.as || col.ref?.[col.ref.length - 1] || (typeof col === 'string' && col)
+      if (col.element.isAssociation) {
+        if (one) this._changeToStreams(col.SELECT.columns, rows[0][name], false, compat)
+        else
+          rows.forEach(row => { this._changeToStreams(col.SELECT.columns, row[name], false, compat) }) 
+      } else if (col.element?.type === 'cds.LargeBinary') {
+        if (one) rows[0][name] = this._stream(rows[0][name])
+        else
+          rows.forEach(row => {
+            row[name] = this._stream(row[name])
+          })        
+      }
     }
-
-    return result
+  } 
+
+  _stream(val) {
+    if (val === null) return null
+    // Buffer.from only applies encoding when the input is a string
+    let raw = typeof val === 'string' ? Buffer.from(val.toString(), 'base64') : val
+    return new Readable({
+      read(size) {
+        if (raw.length === 0) return this.push(null)
+        const chunk = raw.slice(0, size)
+        raw = raw.slice(size)
+        this.push(chunk)
+      }
+    })    
   }
 
   /**
    * Handler for SELECT
    * @type {Handler}
    */
   async onSELECT({ query, data }) {
+    query.SELECT.expand = 'root'
```

Comment:
```suggestion
```

Reply:
@danjoa: We moved `expand = 'root'` from cqn2sql back to onSELECT, because only there we are able to modify it in case of streaming. Otherwise expand is always set to 'root', which leads to building a json, which breaks streaming.